### PR TITLE
146 Create Discussion API Stack

### PR DESCRIPTION
**Pull request by kiikoh**
# This PR creates the completed discussion API stack

### Discussions
```
GET http://localhost:4200/api/v1/auth/discussions/goal/:id
GET http://localhost:4200/api/v1/auth/discussions/topic/:id

PATCH http://localhost:4200/api/v1/auth/discussions/goal/:id
PATCH http://localhost:4200/api/v1/auth/discussions/topic/:id

POST http://localhost:4200/api/v1/auth/discussions/goal/:id
POST http://localhost:4200/api/v1/auth/discussions/topic/:id
```

### Comments
```
POST http://localhost:4200/api/v1/auth/discussions/comment
PATCH http://localhost:4200/api/v1/auth/discussions/comment/:commentId
DELETE http://localhost:4200/api/v1/auth/discussions/comment/:commentId
```

### Ratings
```
POST http://localhost:4200/api/v1/auth/discussions/rating/:commentId
DELETE http://localhost:4200/api/v1/auth/discussions/rating/:commentId
```

### CODE REVIEW

**Comment by kiikoh** [comment]
Intentionally not requesting review just yet as there are still the Swagger API definitions that need to go in before this PR is ready

File: server/routes/apis/discussionRoutes.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    .get( async ( req, res ) => {
+        discussionController.getDiscussionByTopicId( req, res );
+    })
+    // update a discussion based off goal id
+    .patch( async ( req, res ) => { 
+        discussionController.updateDiscussionByTopicId( req, res );
+    })
+    .post(async ( req, res ) => {
+        discussionController.createDiscussionByTopicId( req, res );
+    })
```
Comment (cmarcusdimarco) [inline server/routes/apis/discussionRoutes.js:47]: Comments on these lines seem inverted. Perhaps switching them would be appropriate?


```suggestion
router.route( '/goal/:id' )
    // get a discussion based off goal id
    .get( async ( req, res ) => {
        discussionController.getDiscussionByGoalId( req, res );
    })
    // update a discussion based off goal id
    .patch( async ( req, res ) => { 
        discussionController.updateDiscussionByGoalId( req, res );
    })
    .post(async ( req, res ) => {
        discussionController.createDiscussionByGoalId( req, res );
    })

router.route( '/topic/:id' )
    // get a discussion based off topic id
    .get( async ( req, res ) => {
        discussionController.getDiscussionByTopicId( req, res );
    })
    // update a discussion based off topic id
    .patch( async ( req, res ) => { 
        discussionController.updateDiscussionByTopicId( req, res );
    })
    .post(async ( req, res ) => {
        discussionController.createDiscussionByTopicId( req, res );
    })
```

File: server/routes/apis/discussionRoutes.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    .delete( async (req, res ) => {
+        discussionController.removeRating( req, res )
+    });
+ 
+// leave a comment
+router.route( '/comment' )
+     // enroll an eligible user in a visibile goal
+     .post( async ( req, res ) => {
+        discussionController.createComment( req, res )
+     })
```
Comment (cmarcusdimarco) [inline server/routes/apis/discussionRoutes.js:65]: Comments on lines 51, 55 and 62 should either be removed or updated to reflect current functionality. Looks like these are remnants of the goalRoutes template.

File: server/controller/apis/discussionController.js [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+
+    if(!comment) {
+        const message = ApiMessage.createApiMessage( 404, "Not Found", "Comment not found" );
+        res.set( "x-agora-message-title", "Not Found" );
+        res.set( "x-agora-message-detail", "Comment not found" );
+        return res.status( 404 ).json( message );
+    }
+    
+    res.set( "x-agora-message-title", "Success" );
+    res.set( "x-agora-message-detail", "Returned all goals" );
```
Comment (cmarcusdimarco) [inline server/controller/apis/discussionController.js:286]: API message needs to be updated.
```suggestion
    res.set( "x-agora-message-detail", "Edited comment" );
```

File: server/db/createTablesScript-base-example.sql [hunk 4]
```diff
@@ -571,3 +571,38 @@ GRANT USAGE, SELECT ON SEQUENCE completed_activity_id_seq TO agora;
...
+
+GRANT ALL PRIVILEGES ON TABLE discussions TO agora;
+
+CREATE TABLE IF NOT EXISTS discussion_comments (
+    comment_id SERIAL PRIMARY KEY,
+    parent_id INTEGER, 
+    parent_type discussion_parents,
+    comment_text VARCHAR,
+    creation_date TIMESTAMP DEFAULT current_timestamp,
+  	updated_date TIMESTAMP DEFAULT current_timestamp,
```
Comment (cmarcusdimarco) [inline server/db/createTablesScript-base-example.sql:594]: Is this indentation intended?

File: server/model/comment.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
+function comment() {
+    this.id = -1
+    this.parent_id = -1
+    this.parent_type = ""
+    this.comment_text = ""
+    this.user_id = -1
+    this.created_at = new Date()
+    this.updated_at = new Date()
```
Comment (cmarcusdimarco) [inline server/model/comment.js:8]: We should add terminating semicolons here. Although filled in by JavaScript's ASI, current convention across the other files is to include them, so we should do so to maximize readability and consistency across the codebase.

File: server/model/comment.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+    comment.id = row.comment_id
+    comment.parent_id = row.parent_id
+    comment.parent_type = row.parent_type
+    comment.comment_text = row.comment_text
+    comment.user_id = row.user_id
+    comment.likes = parseInt(row.likes)
+    comment.dislikes = parseInt(row.dislikes)
+    comment.created_at = row.creation_date
+    comment.updated_at = row.updated_date
```
Comment (cmarcusdimarco) [inline server/model/comment.js:28]: Please see lines 2-8.

File: server/model/discussion.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+exports.emptyDiscussion = () => {
+    return new discussion()
+}
+
+exports.ormDiscussion = (row) => {
+    let discussion = exports.emptyDiscussion()
+
+    discussion.parent_id = row.parent_id
+    discussion.parent_type = row.parent_type
+    discussion.discussion_text = row.discussion_text
```
Comment (cmarcusdimarco) [inline server/model/discussion.js:17]: Please see the comments made on model/comment.js re: terminating semicolons.

File: server/controller/apis/discussionController.js [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+        authUserId = req.session.authUser.id;
+    }
+
+    const id = req.params.id
+
+    // get the discussion by goal id
+    let discussion = await discussionService.getDiscussion( "goal", id, authUserId );
+
+    if(!discussion) {
+        const message = new ApiMessage( 404, "Not Found", "Discussion not found" );
```
Comment (cmarcusdimarco) [inline server/controller/apis/discussionController.js:30]: If trying to replicate the use of ApiMessage seen in resourceController.js, this should call ApiMessage's createApiMessage function.
```suggestion
        const message = new ApiMessage.createApiMessage( 404, "Not Found", "Discussion not found" );
```

This applies for all uses of ApiMessage in the file, as testing without this declaration provides the following error:
UnhandledPromiseRejectionWarning: TypeError: ApiMessage is not a constructor

**Review by cmarcusdimarco — CHANGES_REQUESTED** [review]
Excellent work overall. Please see the requested changes and feedback. Some feedback is more of a question than a requested change, but I believe all of it is worth our attention and worth exploring.

Thank you!

File: server/db/createTablesScript-base-example.sql (hunk 4 above)
Comment (kiikoh) [reply server/db/createTablesScript-base-example.sql:594]: It is not intended, but I don't see it showing in my editor either... 
image

File: server/controller/apis/discussionController.js (hunk 8 above)
Comment (kiikoh) [reply server/controller/apis/discussionController.js:30]: I guess because it is not exported, rather confusing

**Review by briangormanly — APPROVED** [review]
This looks good.
I am giving my approval so it can go through once all of @cmarcusdimarco comments are completed.

**Comment by cmarcusdimarco** [comment]
@briangormanly ready when you are for the merge!

**Review by cmarcusdimarco — APPROVED** [review]
Looks like the delete request for resources was overwritten, but in the interest of getting everything within this PR into main and available for everyone, I'm recommending we approve the pull request and create a new issue requesting the addition of the delete resources YAML.

**Pull request merged**